opentelemetry-api/src/opentelemetry/resources/__init__.py

```diff
@@ -11,3 +11,56 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+from typing import Any, Dict, Union
+
+
+class Resource:
+    def __init__(self, labels: Dict[str, str]):
+        """
+        Construct a resource. Direct calling of the
+        constructor is discouraged, as it cannot
+        take advantage of caching and restricts
+        to the type of object that can be returned.
+        """
+        self._labels = labels
+
+    @staticmethod
+    def create(labels: Dict[str, str]) -> "Resource":
+        """
+        create a new resource. This is the recommended
+        method to use to create a new resource.
+        """
+        return Resource(labels)
+
+    def merge(self, other: Union["Resource", None]) -> "Resource":
+        """
+        Perform a merge of the resources, resulting
+        in a union of the resource objects.
+
+        labels that exist in the main Resource take
+        precedence unless the label value is empty.
+        """
+        if other is None:
+            return self
+        if not self._labels:
+            return other
+        merged_labels = self.get_labels().copy()
+        for key, value in other.get_labels().items():
+            if key not in merged_labels or merged_labels[key] == "":
+                merged_labels[key] = value
+        return Resource(merged_labels)
+
+    def get_labels(self) -> Dict[str, str]:
+        """
+        Return the labels associated with this resource.
+
+        get_labels exposes the raw internal dictionary,
+        and as such it is not recommended to copy the
+        result if it is desired to mutate the result.
+        """
+        return self._labels
```

Comment:
Why give `_labels` the underscore treatment if this method just returns it directly?

Reply:
The desire was to match the API specification, which calls for GetLabels.

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-resources.md#getlabels

That said now that I understand that the idea is to provide capabilities rather than specific method names / etc, I can just remove this and make labels public.

Reply:
I switched this over to a property. I was hoping that I could assign values directly and have that satisfy and ABC's property interface, but unfortunately that didn't work as expected.

I've tried to reach a middle ground in the code, which still has a private _labels dict. But it does expose labels in a more Pythonic fashion, while ensuring that any class extending the ABC will still have a labels property.